setup.py

```diff
@@ -1,191 +1,241 @@
-# Copyright (c) 2008-2013 by Enthought, Inc.
+# Copyright (c) 2008-2019 by Enthought, Inc.
 # All rights reserved.
+
 import os
-import re
+import runpy
 import subprocess
-import sys
 
-from setuptools import setup, Extension, find_packages
+import setuptools
 
+# Version information; update this by hand when making a new bugfix or feature
+# release. The actual package version is autogenerated from this information
+# together with information from the version control system, and then injected
+# into the package source.
 MAJOR = 5
 MINOR = 2
 MICRO = 0
-
 IS_RELEASED = False
 
-VERSION = '%d.%d.%d' % (MAJOR, MINOR, MICRO)
+# Templates for version strings.
+RELEASED_VERSION = "{major}.{minor}.{micro}"
+UNRELEASED_VERSION = "{major}.{minor}.{micro}.dev{dev}"
 
+# Paths to the autogenerated version file and the Git directory.
+HERE = os.path.abspath(os.path.dirname(__file__))
+VERSION_FILE = os.path.join(HERE, "traits", "version.py")
+GIT_DIRECTORY = os.path.join(HERE, ".git")
 
-# Return the git revision as a string
-def git_version():
-    def _minimal_ext_cmd(cmd):
-        # construct minimal environment
-        env = {}
-        for k in ['SYSTEMROOT', 'PATH']:
-            v = os.environ.get(k)
-            if v is not None:
-                env[k] = v
-        # LANGUAGE is used on win32
-        env['LANGUAGE'] = 'C'
-        env['LANG'] = 'C'
-        env['LC_ALL'] = 'C'
-        out = subprocess.Popen(
-            cmd, stdout=subprocess.PIPE, env=env,
-        ).communicate()[0]
-        return out
-
-    try:
-        encoded_git_revision = _minimal_ext_cmd(['git', 'rev-parse', 'HEAD'])
-    except OSError:
-        git_revision = "Unknown"
-    else:
-        git_revision = encoded_git_revision.decode("ascii").strip()
-
-    try:
-        out = _minimal_ext_cmd(['git', 'describe', '--tags'])
-    except OSError:
-        out = ''
-
-    git_description = out.decode('ascii').strip()
-    expr = r'.*?\-(?P<count>\d+)-g(?P<hash>[a-fA-F0-9]+)'
-    match = re.match(expr, git_description)
-    if match is None:
-        git_count = '0'
-    else:
-        git_count = match.group('count')
+# Template for the autogenerated version file.
+VERSION_FILE_TEMPLATE = """\
+# Copyright (c) 2008-2019 by Enthought, Inc.
+# All rights reserved.
+
+\"""
```

Comment:
This makes me feel a bit dirty. We're only actually escaping *one* of the three double quotes, but that's enough to avoid this being interpreted as the end of the triple-quoted string.

Reply:
I usually use triple single-quotes when I want to have a string with embedded triple double-quotes.  I think doing that here might improve readability and make you feel less dirty.

Reply:
I'll do that. As a bonus, black leaves that solution unchanged. (I ran the entirety of `setup.py` through black as part of this cleanup.)